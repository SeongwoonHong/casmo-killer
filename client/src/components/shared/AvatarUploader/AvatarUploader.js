import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { validateImg } from '@sharedUtils/inputValidators';

import './AvatarUploader.scss';

class AvatarUploader extends Component {

  constructor(props) {

    super(props);

    this.onImageUpload = this.onImageUpload.bind(this);

  }

  onImageUpload(e) {

    const reader = new FileReader();
    const image = e.target.files[0];

    let message = '';

    reader.onloadend = () => {

      if (image.size > 5000000) {
        message = 'File is too big.';
      } else if (!validateImg(reader.result)) {
        message = 'File format is not supported.';
      }

      this.props.onChange({
        value: message.length === 0 ? reader.result : undefined,
        message
      });

    };

    reader.readAsDataURL(image);

  }

  render() {

    const { className, avatar } = this.props;

    return (
      <div className={ `Avatar-uploader ${className}`}>
        <label>Profile Picture</label>
        <div className="Avatar-uploader__preview">
          <div className="Avatar-uploader__wrapper">
            {
              avatar.value
                ? (
                  <img
                    className="Avatar-uploader__img"
                    alt="user-avatar"
                    src={ avatar.value } />
                )
                : <span>No Image</span>
            }
          </div>
          <input
            type="file"
            accept="image/*"
            id="profilePicture"
            className="Avatar-uploader__input"
            onChange={ this.onImageUpload } />
          <label
            htmlFor="profilePicture"
            className="Avatar-uploader__label">
            <span className="User__form__btn Avatar-uploader__btn">
              Upload New Picture
            </span>
            <span className="Avatar-uploader__text">
              Max 5mb, JPG, or PNG
            </span>
          </label>
        </div>
      </div>
    );
  }

}

AvatarUploader.propTypes = {
  className: PropTypes.string,
  avatar: PropTypes.any,
  onChange: PropTypes.func.isRequired
};

AvatarUploader.defaultProps = {
  className: '',
  avatar: {}
};

export default AvatarUploader;
