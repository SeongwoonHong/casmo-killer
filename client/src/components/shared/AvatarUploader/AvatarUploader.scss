.Avatar-uploader {
    &__preview {
        @include flex-box(flex-start, flex-end);
        margin-top: $user-box-padding;
    }
    &__wrapper {
        @include flex-box(center);
        @include rectanglify(100px);
        background-color: palette(Pink, 50);
        border: 1px solid palette(Pink, 100);
        border-radius: 50%;
        color: palette(Pink, 100);
    }
    &__img {
        @include rectanglify(100px);
        border-radius: 50%;
    }
    &__input {
        position: absolute;
        z-index: -88;
    }
    &__label {
        margin-left: $user-box-padding;
        text-align: center;
    }
    &__btn {
        color: palette(Grey, 50);
        margin-bottom: $user-box-padding - 10px;
    }
    &__text {
        color: palette(Grey, 600);
    }
}
