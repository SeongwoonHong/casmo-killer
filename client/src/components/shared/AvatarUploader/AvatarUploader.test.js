import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import AvatarUploader from './AvatarUploader';

describe('<AvatarUploader />', () => {

  const props = {
    className: 'testing',
    avatar: {
      value: 'testing'
    },
    onChange: () => {},
    onError: () => {}
  };

  it('renders the component consistently', () => {

    const component = shallow(
      <AvatarUploader { ...props } />
    );

    expect(component).toMatchSnapshot();

  });

  it('fires onChange event handler', () => {

    const spy = sinon.spy(AvatarUploader.prototype, 'onImageUpload');

    const target = {
      files: [
        new Blob(['testing'], {
          type: 'img/*'
        })
      ]
    };

    const component = shallow(<AvatarUploader{ ...props } />);

    component.find('input').simulate('change', { target });

    expect(spy.calledWith({ target })).toEqual(true);

  });

});
