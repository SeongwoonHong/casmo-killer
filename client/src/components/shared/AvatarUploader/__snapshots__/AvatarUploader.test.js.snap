// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarUploader /> renders the component consistently 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AvatarUploader
    avatar={
        Object {
            "value": "testing",
          }
    }
    className="testing"
    onChange={[Function]}
    onError={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <label>
          Profile Picture
</label>,
        <div
          className="Avatar-uploader__preview"
>
          <div
                    className="Avatar-uploader__wrapper"
          >
                    <img
                              alt="user-avatar"
                              className="Avatar-uploader__img"
                              src="testing"
                    />
          </div>
          <input
                    accept="image/*"
                    className="Avatar-uploader__input"
                    id="profilePicture"
                    onChange={[Function]}
                    type="file"
          />
          <label
                    className="Avatar-uploader__label"
                    htmlFor="profilePicture"
          >
                    <span
                              className="User__form__btn Avatar-uploader__btn"
                    >
                              Upload New Picture
                    </span>
                    <span
                              className="Avatar-uploader__text"
                    >
                              Max 5mb, JPG, or PNG
                    </span>
          </label>
</div>,
      ],
      "className": "Avatar-uploader testing",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Profile Picture",
        },
        "ref": null,
        "rendered": "Profile Picture",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="Avatar-uploader__wrapper"
>
              <img
                            alt="user-avatar"
                            className="Avatar-uploader__img"
                            src="testing"
              />
</div>,
            <input
              accept="image/*"
              className="Avatar-uploader__input"
              id="profilePicture"
              onChange={[Function]}
              type="file"
/>,
            <label
              className="Avatar-uploader__label"
              htmlFor="profilePicture"
>
              <span
                            className="User__form__btn Avatar-uploader__btn"
              >
                            Upload New Picture
              </span>
              <span
                            className="Avatar-uploader__text"
              >
                            Max 5mb, JPG, or PNG
              </span>
</label>,
          ],
          "className": "Avatar-uploader__preview",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="user-avatar"
                className="Avatar-uploader__img"
                src="testing"
/>,
              "className": "Avatar-uploader__wrapper",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "user-avatar",
                "className": "Avatar-uploader__img",
                "src": "testing",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "accept": "image/*",
              "className": "Avatar-uploader__input",
              "id": "profilePicture",
              "onChange": [Function],
              "type": "file",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span
                  className="User__form__btn Avatar-uploader__btn"
>
                  Upload New Picture
</span>,
                <span
                  className="Avatar-uploader__text"
>
                  Max 5mb, JPG, or PNG
</span>,
              ],
              "className": "Avatar-uploader__label",
              "htmlFor": "profilePicture",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Upload New Picture",
                  "className": "User__form__btn Avatar-uploader__btn",
                },
                "ref": null,
                "rendered": "Upload New Picture",
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Max 5mb, JPG, or PNG",
                  "className": "Avatar-uploader__text",
                },
                "ref": null,
                "rendered": "Max 5mb, JPG, or PNG",
                "type": "span",
              },
            ],
            "type": "label",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label>
            Profile Picture
</label>,
          <div
            className="Avatar-uploader__preview"
>
            <div
                        className="Avatar-uploader__wrapper"
            >
                        <img
                                    alt="user-avatar"
                                    className="Avatar-uploader__img"
                                    src="testing"
                        />
            </div>
            <input
                        accept="image/*"
                        className="Avatar-uploader__input"
                        id="profilePicture"
                        onChange={[Function]}
                        type="file"
            />
            <label
                        className="Avatar-uploader__label"
                        htmlFor="profilePicture"
            >
                        <span
                                    className="User__form__btn Avatar-uploader__btn"
                        >
                                    Upload New Picture
                        </span>
                        <span
                                    className="Avatar-uploader__text"
                        >
                                    Max 5mb, JPG, or PNG
                        </span>
            </label>
</div>,
        ],
        "className": "Avatar-uploader testing",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Profile Picture",
          },
          "ref": null,
          "rendered": "Profile Picture",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="Avatar-uploader__wrapper"
>
                <img
                                alt="user-avatar"
                                className="Avatar-uploader__img"
                                src="testing"
                />
</div>,
              <input
                accept="image/*"
                className="Avatar-uploader__input"
                id="profilePicture"
                onChange={[Function]}
                type="file"
/>,
              <label
                className="Avatar-uploader__label"
                htmlFor="profilePicture"
>
                <span
                                className="User__form__btn Avatar-uploader__btn"
                >
                                Upload New Picture
                </span>
                <span
                                className="Avatar-uploader__text"
                >
                                Max 5mb, JPG, or PNG
                </span>
</label>,
            ],
            "className": "Avatar-uploader__preview",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  alt="user-avatar"
                  className="Avatar-uploader__img"
                  src="testing"
/>,
                "className": "Avatar-uploader__wrapper",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "user-avatar",
                  "className": "Avatar-uploader__img",
                  "src": "testing",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "accept": "image/*",
                "className": "Avatar-uploader__input",
                "id": "profilePicture",
                "onChange": [Function],
                "type": "file",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <span
                    className="User__form__btn Avatar-uploader__btn"
>
                    Upload New Picture
</span>,
                  <span
                    className="Avatar-uploader__text"
>
                    Max 5mb, JPG, or PNG
</span>,
                ],
                "className": "Avatar-uploader__label",
                "htmlFor": "profilePicture",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Upload New Picture",
                    "className": "User__form__btn Avatar-uploader__btn",
                  },
                  "ref": null,
                  "rendered": "Upload New Picture",
                  "type": "span",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Max 5mb, JPG, or PNG",
                    "className": "Avatar-uploader__text",
                  },
                  "ref": null,
                  "rendered": "Max 5mb, JPG, or PNG",
                  "type": "span",
                },
              ],
              "type": "label",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
